### ISSUE TITLE
Type safe field reference for values table

### ISSUE THREAD

**Issue by baev** [jOOQ/jOOQ#10766]
We are using a lot of multi-insert queries that like the following:

```java

Row2<A, B>[] rows = ...;
Table<Record2<A, B>> data = DSL.values(rows);
ctx
  .insertInto(T)
  .columns(T.C1, T.C2, T.C3)
  .select(DSL
    .select(
      val(someConstC1),
      data.field(0, SQLDataType.A),
      data.field(1, SQLDataType.B)
    )
    .from(data)
  );
```

And it just annoying to have this field numbers and data types all over the query (those fields are used in join and where conditions as well). So as workaround I found that we can use `newRecord` and reference data's fields in type-safe way:

```java
Row2<A, B>[] rows = ...;
Table<Record2<A, B>> data = DSL.values(rows);
Record2<A, B> dataRec = data.newRecord();
ctx
  .insertInto(T)
  .columns(T.C1, T.C2, T.C3)
  .select(DSL
    .select(
      val(someConstC1),
      dataRec.field1(),
      dataRec.field2(),
    )
    .from(data)
  );
```

this much more safe in terms of types but just feels oddly that I have to call a `newRecord` method. So maybe there could be some sort of other way to have type-safe fields reference without creating a new record? If not, could that be added? Apparently that would require to generate another 23+ classes with `fieldN` methods...

Am I missing something? 

### Versions:

- jOOQ: 3.14.0
- Java: 11.0.2
- Database (include vendor): PostgreSQL 12
- OS: macos
- JDBC Driver (include name if inofficial driver):

**Comment by lukaseder** [comment]
Thanks a lot for your question.

Cunning workaround with the `newRecord()` call, to get type safe access to the `field[N]` methods! Given what you're doing here, I don't even think it's that bad! 😀

There are different ways to think about this operation. For example, you could also cross join the 1x1 table `someConstC1` with your `data` table, but that would remove type safety. Alternatively, you could, of course, construct `Row3<C1Type, A, B>` rows instead?

If removing type safety is fine, then this would be much simpler:

```java
Row2<A, B>[] rows = ...;
Table<Record2<A, B>> data = DSL.values(rows);
ctx
  .insertInto(T)
  .columns(Arrays.asList(T.C1, T.C2, T.C3)) // Prevent applying type safe API, or even T.fields()
  .select(DSL
    .select(val(someConstC1))
    .select(data.fields())
    .from(data)
  );
```

Fact is, with Java, a `Table<Record2<A, B>>` type cannot really reflect on its columns, because to the `Table<R>` type doesn't have access to the specific type bound of `<R>`. It would be possible with 22*23/2 generic methods (not 22), such as 

```java
static <T> Field<T> field1(Table<? extends Record1<T>>);
static <T> Field<T> field1(Table<? extends Record2<T, ?>);
static <T> Field<T> field2(Table<? extends Record2<?, T>);
static <T> Field<T> field1(Table<? extends Record3<T, ?, ?>);
static <T> Field<T> field2(Table<? extends Record3<?, T, ?>);
static <T> Field<T> field3(Table<? extends Record3<?, ?, T>);
...
```

But the use-case seems too marginal to justify this extreme overhead.

So, in short, I don't have a satisfying answer to this question yet. I will close it as "answered", hoping that maybe, eventually, something might come to mind.

**Comment by baev** [comment]
Thanks for the answer! 

> Alternatively, you could, of course, construct Row3<C1Type, A, B> rows instead?

Trying to keep X in RowX as low as possible. Increasing number of columns in row require adding additional utility code (that I mentioned in #6604 ). Plus, the typical use-case contains cross join and conditions (and probably mix it with `val` as well)

```java
Table<Record2<A, B>> data = ...
Record2<A, B> dataRec = data.newRecord();
ctx
  .insertInto(T)
  .columns(T.C1, T.C2, T.C3)
  .select(DSL
    .select(
      X.C1,
      dataRec.field1(),
      dataRec.field2(),
    )
    .from(data, X)
    .where(X.Y.eq(dataFields.field1())) // or some other X condition
  );
```


> Fact is, with Java, a `Table<Record2<A, B>>` type cannot really reflect on its columns, because to the `Table<R>` type doesn't have access to the specific type bound of `<R>`. It would be possible with 22*23/2 generic methods (not 22), such as

the solution I have in mind is to create `TypedTableFields` interfaces and add the `Table#getTableFields` method:

```java
Table<Record2<A, B>> data = ...
TypedTableFields<A, B> dataFields = data.getTypedFields();
ctx
  .insertInto(T)
  .columns(T.C1, T.C2, T.C3)
  .select(DSL
    .select(
      val(someConstC1),
      dataFields.field1(),
      dataFields.field2(),
    )
    .from(data)
  );
```

**Comment by lukaseder** [comment]
> the solution I have in mind is to create `TypedTableFieldsN` interfaces and add the `Table#getTableFields` method:

I invite you to do the exercise and design these types. How could they possibly access the `Table<R extends Record>`'s `<R>` type variable content?

Example, I have a type:

```java
interface X extends Record {}
```

Now I create a reference `Table<X>`. What would your `TypedTableFieldsN` interface do with that?

**Comment by baev** [comment]
the easiest way is to add TypedTableFieldsN to RecordN, and then access RecordN the same way to done for `newRecord()` method

**Comment by lukaseder** [comment]
OK, so how would that work:

```java
Table<Record2<A, B>> data = ...
TypedTableFields2<A, B> dataFields = data.getTypedFields();
```

What's the signature of `Table<R>.getTypedFields()`?

**Comment by baev** [comment]
Ah, I see; so, I guess `Record2<A, B> dataRec = data.newRecord()` is way to go, then, thanks

**Comment by lukaseder** [comment]
Yes, I'm afraid it is. If only Java had higher kinded generics, or Kotlin's extension functions...
